### PR TITLE
Test/mkldnn batch norm op

### PR DESCRIPTION
**Pull request by azai91**
## Description ##
Add tests for MKLDNN batch norm operator The creates ndarray inputs and then runs it separately through the native CPU BN operator and MKLDNN BN operator. Creating the BN test requires a different helper as the BN operation writes to the input, so identical arrays of equal value must be created and then ran through the native / MKLDNN operators.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Add tests for MKLDNN BN operator

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by ankkhedia** [comment]
@azai91 Thanks for your contribution!

@mxnet-label-bot [pr-awaiting-review , MKLDNN]

@mseth10 Could you please take a look into PR?

**Comment by pengzhao-intel** [comment]
@TaoLv please take a look if there's conflict with our latest upgrade for the convolution.

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc [hunk 1]
```diff
@@ -197,32 +198,20 @@ MKLDNNDeconvForward::MKLDNNDeconvForward(const DeconvolutionParam& param,
...
                                         const std::vector<NDArray> &out_data) {
-  auto data_mem = in_data[deconv::kData].GetMKLDNNDataReorder(
+  auto data_mem = in_data.GetMKLDNNDataReorder(
       fwd_pd.diff_dst_primitive_desc());
-  NDArray weight = in_data[deconv::kWeight];
   const mkldnn::memory *weight_mem;
   if (ctx.is_train) {
-    // TODO(zhengda) kvstore doesn't handle MKLDNN correctly. Let's reorder it
-    // to the default format for now.
-    if (weight.IsMKLDNNData())
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Why this is removed?

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc [hunk 2]
```diff
@@ -197,32 +198,20 @@ MKLDNNDeconvForward::MKLDNNDeconvForward(const DeconvolutionParam& param,
...
-      // This asks the engine to reorder data after the weight array is used.
-      weight.Reorder2DefaultAsync();
     weight_mem = GetWeights(weight, fwd_pd.weights_primitive_desc(), param.num_group);
   } else {
     // For inference, we want to reorder the weight array so we don't need to
     // reorder data every time.
-    if (weight.IsDefaultData()) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Ditto

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc [hunk 3]
```diff
@@ -197,32 +198,20 @@ MKLDNNDeconvForward::MKLDNNDeconvForward(const DeconvolutionParam& param,
...
-      weight_mem = GetWeights(weight, fwd_pd.weights_primitive_desc(), param.num_group);
-      // We also need to modify the layout on the original weight array. The
-      // data conversion happens after the weight array is used.
-      weight.MKLDNNDataReorderAsync(fwd_pd.weights_primitive_desc());
-    } else {
-      weight_mem = weight.GetMKLDNNData();
-      CHECK(weight_mem->get_primitive_desc() == fwd_pd.weights_primitive_desc());
-    }
+    weight_mem = weight.GetMKLDNNData();
+    CHECK(weight_mem->get_primitive_desc() == fwd_pd.weights_primitive_desc());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Do we really need this assumption?

File: tests/cpp/include/test_mkldnn.h [hunk 4]
```diff
@@ -240,6 +244,38 @@ enum ArrayTypes {
   All = 8191,
 };
 
+
+inline NDArray CreateKernelNDArray(TShape kernel, int num_filters, TShape input,
+    bool is_deconv = false) {
+  CHECK(kernel.ndim() == 2) << "mkldnn only supports 2d filters on 4d inputs";
```
Comment (TaoLv) [inline tests/cpp/include/test_mkldnn.h]: CHECK_EQ

File: tests/cpp/operator/mkldnn_operator_test.cc [hunk 5]
```diff
@@ -213,6 +215,121 @@ OpAttrs GetLRNBackwardsOp() {
...
+  attrs.attrs.dict.insert({"num_filter" , std::to_string(num_filters)});
+  attrs.attrs.dict.insert({"stride" , CreateShapeString(stride, dim)});
+  attrs.attrs.dict.insert({"pad" , CreateShapeString(pad, dim)});
+  attrs.attrs.op->attr_parser(&attrs.attrs);
+  return attrs;
+}
+
+OpAttrs GetBNOp() {
+  OpAttrs attrs;
+  attrs.attrs.op = Op::Get("LRN");
```
Comment (TaoLv) [inline tests/cpp/operator/mkldnn_operator_test.cc]: BN?

File: tests/cpp/operator/mkldnn_operator_test.cc [hunk 6]
```diff
@@ -213,6 +215,121 @@ OpAttrs GetLRNBackwardsOp() {
...
+      ArrayTypes::NormalReshaped |
+      ArrayTypes::MKLDNNReshaped;
+  attrs.output_types = ArrayTypes::Normal |
+      ArrayTypes::MKLDNN |
+      ArrayTypes::NormalReshaped |
+      ArrayTypes::MKLDNNReshaped;
+  return attrs;
+}
+
+OpAttrs GetBNBackwardsOp() {
```
Comment (TaoLv) [inline tests/cpp/operator/mkldnn_operator_test.cc]: Backwards -> Backward

File: tests/cpp/operator/mkldnn_operator_test.cc [hunk 7]
```diff
@@ -213,6 +215,121 @@ OpAttrs GetLRNBackwardsOp() {
...
+  OpAttrs attrs;
+  attrs.attrs.op = Op::Get("_backward_LRN");
```
Comment (TaoLv) [inline tests/cpp/operator/mkldnn_operator_test.cc]: BN

File: tests/cpp/operator/mkldnn_operator_test.cc [hunk 8]
```diff
@@ -552,6 +673,201 @@ void TestOpEx(const OpAttrs &forward_attrs, const OpAttrs &backwards_attrs) {
...
+        std::cout << "Backwards: ";
+        PrintVerifyMsg(out_arrs[0][output_i], tmp_output);
+        Imperative::Get()->InvokeOp(
+            Context(), backwards_attrs.attrs, backwards_input, backwards_outputs,
+            back_req, DispatchMode::kFCompute, mxnet::OpStatePtr());
+        Imperative::Get()->InvokeOp(
+            Context(), backwards_attrs.attrs, backwards_input, backwards_ex_outputs,
+            back_req, DispatchMode::kFComputeEx, mxnet::OpStatePtr());
+        Engine::Get()->WaitForAll();
+        AssertEqual(backwards_outputs, backwards_ex_outputs);
```
Comment (TaoLv) [inline tests/cpp/operator/mkldnn_operator_test.cc]: What's the tolerance for this check?

**Comment by azai91** [comment]
@TaoLv please review

**Comment by Vikas89** [comment]
@azai91 It will be helpful if you can please add more details to description explaining why tests where failing earlier, and how did you fix it.

**Comment by lupesko** [comment]
@mseth10 @anirudh2290 pulling you guys in for a review...

**Comment by azai91** [comment]
@Vikas89 added

**Comment by TaoLv** [comment]
Is it possible to separate operator change and unit test into two PRs?

**Comment by azai91** [comment]
@TaoLv https://github.com/apache/incubator-mxnet/pull/13625

File: src/operator/nn/batch_norm.cc [hunk 9]
```diff
@@ -380,12 +380,20 @@ static bool BatchNormType(const nnvm::NodeAttrs& attrs,
...
+    const BatchNormParam &param) {
+  TShape shape = inputs[0].shape();
+  bool params_valid = shape.ndim() == 4
       && param.axis == mxnet::op::batchnorm::DEFAULT_AXIS
       && shape[param.axis] % 8 == 0
       && !mxnet::op::batchnorm::disable_mkl;
+  bool inputs_valid = SupportMKLDNN(inputs[0]);
+  for (size_t i = 1; i < inputs.size(); i++) {
+    if (inputs[i].IsMKLDNNData()) {
+      inputs_valid = false;
```
Comment (anirudh2290) [inline src/operator/nn/batch_norm.cc]: inputs valid is set to false if it is in MKLDNN format ?

**Comment by mseth10** [comment]
@azai91 can you please resolve merge conflicts.

**Comment by sandeep-krishnamurthy** [comment]
@azai91 - can you please rebase?

**Comment by anirudhacharya** [comment]
@azai91 please rebase and resolve merge conflicts

**Comment by azai91** [comment]
closing since this is a duplicate of https://github.com/apache/incubator-mxnet/pull/13625
